src/Conversion/KrnlToAffine/KrnlMatmul.cpp

```diff
@@ -372,13 +372,7 @@ class KrnlMatmulLowering : public ConversionPattern {
     assert(BUFFER_ALIGN >= gDefaultAllocAlign &&
            "alignment of buffers cannot be smaller than the default alignment "
            "(which is set for SIMD correctness");
-    // TODO: alloca is good as it help simplify away this data structures (as it
-    // is only used as local temp, basically extensions of registers). However,
-    // there might be issues with non-removed alloca when they are not in the
-    // innermost loop. Still think its worth it having alloca as we want
-    // eventually all the refs to alloca to be register/spill access, not memory
-    // load/stores.
-    Value TmpProd = create.mem.alignedAlloca(CTmpType, BUFFER_ALIGN);
+    Value TmpProd = create.mem.alignedAlloc(CTmpType, BUFFER_ALIGN);
```

Comment:
If it is a good practice to use `alignedAlloca` for scalar type, should we add a verification in `create.mem.alignedAlloca` that checks type would be scalar. By that way, we can avoid the situation where users call `create.mem.alignedAlloca` for non-scalar type.

Reply:
> I remember we encounted segfaults because of alloca in the past and added the hoisting pass to move it outside loops. It's a pity to hear that that pass does not work with scope_alloca.

`scope_alloca` stop the migration of `alloc` and `alloca` out of that scope. It is important for parallel regions, as having it stay in the parallel region means that each thread will get its own. So that works fine.

Now I also manually moved the case with no parallel out of the loops, just to make sure that is fine. Would you like me to check if it's really needed? That way we could have the call always inside, and if parallel is there, its blocked and otherwise it is not?

In one case, I really need it outside manually, because it was inside of an inner-loop if-then-else. So by having it outside of the if-then-else (allocating in all case), it allows it to migrate to the outside.

Reply:
Left for another PR.